gemma-core/src/test/java/ubic/gemma/core/loader/expression/DataUpdater2Test.java

```diff
@@ -0,0 +1,173 @@
+        }
+
+        @Bean
+        public VectorMergingService vectorMergingService() {
+            return mock( VectorMergingService.class );
+        }
+
+        @Bean
+        public RawAndProcessedExpressionDataVectorService rawAndProcessedExpressionDataVectorService() {
+            return mock( RawAndProcessedExpressionDataVectorService.class );
+        }
+    }
+
+    @Autowired
+    private DataUpdater dataUpdater;
+
+    @Autowired
+    private PreprocessorService preprocessorService;
+
+    @Autowired
+    private ExpressionExperimentService expressionExperimentService;
+
+    @Autowired
+    private ArrayDesignService arrayDesignService;
+
+    @Autowired
+    private BioAssayDimensionService bioAssayDimensionService;
+
+    private ExpressionExperiment ee;
+    private ArrayDesign ad;
+    private DoubleMatrix<String, String> rpkmMatrix, countMatrix;
+
+    @Before
+    public void setUp() throws IOException {
+        rpkmMatrix = new DoubleMatrixReader().read( getClass().getResourceAsStream( "/data/loader/expression/flatfileload/GSE19166_expression_count.test.txt" ) );
+        countMatrix = new DoubleMatrixReader().read( getClass().getResourceAsStream( "/data/loader/expression/flatfileload/GSE19166_expression_RPKM.test.txt" ) );
+        ee = new ExpressionExperiment();
+        for ( String col : countMatrix.getColNames() ) {
+            BioAssay ba = BioAssay.Factory.newInstance( col );
+            BioMaterial bm = BioMaterial.Factory.newInstance( col );
+            bm.setBioAssaysUsedIn( Collections.singleton( ba ) );
+            ba.setSampleUsed( bm );
+            ee.getBioAssays().add( ba );
+        }
+        ad = new ArrayDesign();
+        ad.setId( 1L );
+        for ( String row : countMatrix.getRowNames() ) {
+            ad.getCompositeSequences().add( CompositeSequence.Factory.newInstance( row, ad ) );
+        }
+    }
+
+    @Test
+    public void testSkipRecomputeLog2cpm() {
+        when( expressionExperimentService.thawLite( ee ) ).thenReturn( ee );
+        when( expressionExperimentService.addRawVectors( same( ee ), any() ) ).thenReturn( ee );
+        when( expressionExperimentService.replaceRawVectors( same( ee ), any() ) ).thenReturn( ee );
+        when( arrayDesignService.thaw( ad ) ).thenReturn( ad );
+        when( expressionExperimentService.getArrayDesignsUsed( ee ) ).thenReturn( Collections.singleton( ad ) );
+        when( bioAssayDimensionService.findOrCreate( any() ) ).thenAnswer( a -> a.getArgument( 0 ) );
+        dataUpdater.addCountData( ee, ad, countMatrix, rpkmMatrix, 31, false, false, false );
+        verify( preprocessorService ).process( ee );
+        verify( expressionExperimentService, times( 2 ) ).addRawVectors( same( ee ), any() );
+        dataUpdater.addCountData( ee, ad, countMatrix, rpkmMatrix, 30, false, false, true );
+        verify( expressionExperimentService ).replaceRawVectors( same( ee ), any() );
+        verifyNoMoreInteractions( preprocessorService );
```

Comment:
@ppavlidis I've added the unit test here. It checks that the preprocessor service is only invoked once.

I'll add a test to see how differing library sizes are handled. Let me know if you want other cases covered.